crates/storage/provider/src/transaction/mod.rs

```diff
@@ -0,0 +1,764 @@
+        let hashed = storages.into_iter().fold(BTreeMap::new(), |mut map, (address, storage)| {
+            let storage = storage.into_iter().fold(BTreeMap::new(), |mut map, (key, value)| {
+                map.insert(keccak256(key), value);
+                map
+            });
+            map.insert(keccak256(address), storage);
+            map
+        });
+
+        let mut hashed_storage = self.cursor_dup_write::<tables::HashedStorage>()?;
+        // Hash the address and key and apply them to HashedStorage (if Storage is None
+        // just remove it);
+        hashed.into_iter().try_for_each(|(address, storage)| {
+            storage.into_iter().try_for_each(|(key, value)| -> Result<(), TransactionError> {
+                if hashed_storage
+                    .seek_by_key_subkey(address, key)?
+                    .filter(|entry| entry.key == key)
+                    .is_some()
+                {
+                    hashed_storage.delete_current()?;
+                }
+
+                if value != U256::ZERO {
+                    hashed_storage.upsert(address, StorageEntry { key, value })?;
+                }
+                Ok(())
+            })
+        })?;
+        Ok(())
+    }
+
+    /// Iterate over account changesets and return all account address that were changed.
+    pub fn get_addresses_of_changed_accounts(
+        &self,
+        from: TransitionId,
+        to: TransitionId,
+    ) -> Result<BTreeSet<Address>, TransactionError> {
+        Ok(self
+            .cursor_read::<tables::AccountChangeSet>()?
+            .walk_range(from..to)?
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            // fold all account to one set of changed accounts
+            .fold(BTreeSet::new(), |mut accounts: BTreeSet<Address>, (_, account_before)| {
+                accounts.insert(account_before.address);
+                accounts
+            }))
+    }
+
+    /// Get plainstate account from iterator
+    pub fn get_plainstate_accounts(
+        &self,
+        iter: impl IntoIterator<Item = Address>,
+    ) -> Result<Vec<(Address, Option<Account>)>, TransactionError> {
+        let mut plain_accounts = self.cursor_read::<tables::PlainAccountState>()?;
+        Ok(iter
+            .into_iter()
+            .map(|address| plain_accounts.seek_exact(address).map(|a| (address, a.map(|(_, v)| v))))
+            .collect::<Result<Vec<_>, _>>()?)
+    }
+
+    /// iterate over accounts and insert them to hashing table
+    pub fn insert_account_for_hashing(
+        &self,
+        accounts: impl IntoIterator<Item = (Address, Option<Account>)>,
+    ) -> Result<(), TransactionError> {
+        let mut hashed_accounts = self.cursor_write::<tables::HashedAccount>()?;
+
+        let hashes_accounts = accounts.into_iter().fold(
+            BTreeMap::new(),
+            |mut map: BTreeMap<H256, Option<Account>>, (address, account)| {
+                map.insert(keccak256(address), account);
```

Comment:
@onbjerg do you mean introducing something like a `HashedAddress::from(src: &Address)` type or something? And make sure it's always constructed corretly?

Reply:
No, just that it does not make sense for `Transaction` to do hashing, we should just provide it some type where this is already done. If it's a `HashedAddress` or a `BTreeMap<H256, Account>` or whatever doesn't matter much to me.

Reply:
Just to say my view on that topic, this Transaction is not the same thing as database transactions but it is a provider transaction. There is a difference between what they represent and I see them as:
* db transaction is low level and do anything with tables
* provider transaction (Both SharedDatabase and ContinuousTransaction) are on higher level and they know about types/data/interconnections between them. So transformation related to data seems good to be placed here.

For the pattern to use to organize this better, I am not sure, some time needs to be spent on experimenting with it but I am focused on delivering the Tree atm. More on it here: https://github.com/paradigmxyz/reth/pull/1474#discussion_r1112928290